numba/core/utils.py

```diff
@@ -751,3 +751,13 @@ def get_hashable_key(value):
         return id(value)
     else:
         return value
+
+
+class threadsafe_cached_property(functools.cached_property):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._lock = threading.RLock()
+
+    def __get__(self, *args, **kwargs):
+        with self._lock:
+            return super().__get__(*args, **kwargs)
```

Comment:
This change is needed due to py3.12 https://github.com/python/cpython/pull/101890 removal of the builtin lock. The reason for the removal being all instances of the class that use `cached_property` share the same lock. For Numba usecase, this `cached_property` class is mainly used for the compiler which is already under the compiler_lock, and the rare situation that some lazy properties are used outside of the compiler pipeline.

The changes here is simple and gets the job done.

Reply:
Thanks for checking @sklam